Src/Newtonsoft.Json/JsonTextWriter.Async.cs

```diff
@@ -0,0 +1,1196 @@
+        {
+            return SafeAsync ? WriteIntegerValueAsync(value, cancellationToken) : base.WriteValueAsync(value, cancellationToken);
+        }
+
+        /// <summary>
+        /// Asynchronously writes a <see cref="Nullable{T}"/> of <see cref="byte"/> value.
+        /// </summary>
+        /// <param name="value">The <see cref="Nullable{T}"/> of <see cref="byte"/> value to write.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous operation.</returns>
+        /// <remarks>Derived classes must override this method to get asynchronous behaviour. Otherwise it will
+        /// execute synchronously, returning an already-completed task.</remarks>
+        public override Task WriteValueAsync(byte? value, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return SafeAsync ? DoWriteValueAsync(value, cancellationToken) : base.WriteValueAsync(value, cancellationToken);
+        }
+
+        internal Task DoWriteValueAsync(byte? value, CancellationToken cancellationToken)
+        {
+            return value == null ? DoWriteNullAsync(cancellationToken) : WriteIntegerValueAsync(value.GetValueOrDefault(), cancellationToken);
+        }
+
+        /// <summary>
+        /// Asynchronously writes a <see cref="byte"/>[] value.
+        /// </summary>
+        /// <param name="value">The <see cref="byte"/>[] value to write.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>A <see cref="Task"/> that represents the asynchronous operation.</returns>
+        /// <remarks>Derived classes must override this method to get asynchronous behaviour. Otherwise it will
+        /// execute synchronously, returning an already-completed task.</remarks>
+        public override Task WriteValueAsync(byte[] value, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return SafeAsync ? (value == null ? WriteNullAsync(cancellationToken) : WriteValueNonNullAsync(value, cancellationToken)) : base.WriteValueAsync(value, cancellationToken);
+        }
+
+        internal async Task WriteValueNonNullAsync(byte[] value, CancellationToken cancellationToken)
+        {
+            await InternalWriteValueAsync(JsonToken.Bytes, cancellationToken).ConfigureAwait(false);
+            await _writer.WriteAsync(_quoteChar, cancellationToken).ConfigureAwait(false);
+            await Base64Encoder.EncodeAsync(value, 0, value.Length, cancellationToken).ConfigureAwait(false);
+            await Base64Encoder.FlushAsync(cancellationToken).ConfigureAwait(false);
+            await _writer.WriteAsync(_quoteChar, cancellationToken).ConfigureAwait(false);
```

Comment:
Not really compared to those that loop (or indeed other such code elsewhere like `Stream.CopyToAsync()`) when one considers that the loop could cycle through many times.